File: mgyoutube-multitech/api-node/src/modules/defaultusermoduleimpl.ts
import { User } from '../apimodel/user';
import { UserCredential } from '../apimodel/usercredential';
import { UserNotFoundException } from './usernotfoundexception';
import { UserModule } from './usermodule';
import { UserDataRepo } from '../repos/userdatarepo';


export class DefaultUserModuleImpl implements UserModule {
    userDataRepo: UserDataRepo;

    constructor(aUserDataRepo: UserDataRepo) {
        this.setUserDataRepo(aUserDataRepo);
    }

    setUserDataRepo(udr: UserDataRepo): void {
        this.userDataRepo = udr;
    }

    async authUser(userCredential: UserCredential): Promise<User | undefined> {
        var user: User | undefined;

        try {
            user = await this.userDataRepo.getUserByUsername(userCredential.username)

            if (user != null) {
                if (user.password == null) {
                    console.log("user.password is None")
                    user = undefined
                }
                else if (userCredential.password == null) {
                    console.warn("userCredential.password is None")
                    user = undefined
                }
                else if (user.password != userCredential.password) {
                    console.warn("user.password(", user.password,
                        ") != userCredential.password(", userCredential.password, ")")
                    user = undefined
                }
                else {
                    console.log("password match!")
                    // password match!
                }
            }
            else {
                console.warn("getUserByUsername(", userCredential.username, ") returned None")
            }
        }
        catch (e) {
            console.error(e)
            user = undefined
        }

        return user
    }

    async createUpdateUser(user: User): Promise<User | undefined> {
        const existingUser = await this.userDataRepo.getUserByUsername(user.username);
        if (existingUser !== undefined) {
            if (user.userId === undefined) {
                user.userId = existingUser.userId;
            }
            console.log("createUpdateUser is updating existing user ", existingUser, " to be ", user);
            return this.updateUser(existingUser.userId, user);
        }
        console.log("createUpdateUser is creating new user ", user);
        return this.createUser(user);
    }

    async createUser(user: User): Promise<User | undefined> {
        if (user.userId === undefined) {
            const createdUser = this.userDataRepo.addUser(user)
            //if (createdUser !== undefined) {
            return createdUser
            //}
        }
        return undefined
    }

    async updateUser(userId: string, user: User): Promise<User | undefined> {
        return this.userDataRepo.replaceUser(userId, user);
    }

    async getUser(username: string): Promise<User | undefined> {
        try {
            return this.userDataRepo.getUserByUsername(username)
        }
        catch (e) {
            console.error(e)
            return undefined
        }
    }

    async removeUser(username: string): Promise<User | undefined> {
        const user = await this.getUser(username);

        if (user && user.userId) {
            await this.userDataRepo.removeUser(user);
        }

        return user;
    }

    async addUpdateChildToParent(parentUsername: string, childUser: User): Promise<User | undefined> {

        const parentUser = await this.getUser(parentUsername);
        if (parentUser === undefined) {
            throw new UserNotFoundException(parentUsername);
        }

        const existingChildUser = await this.getUser(childUser.username);
        if (existingChildUser === undefined) {
            console.log("addUpdateChildToParent, creating child ", childUser);
            const createdChildUser = await this.createUser(childUser);
            if (createdChildUser === undefined) {
                console.error("addUpdateChildToParent, creating child ", childUser, " failed ");
            }
            else {
                await this.userDataRepo.addChildToParent(parentUser.userId, createdChildUser.userId);
            }
            return createdChildUser;
        } else {
            console.log("addUpdateChildToParent, updating child ", childUser, " as ", existingChildUser.userId);
            this.userDataRepo.addChildToParent(parentUser.userId, existingChildUser.userId);
            return this.updateUser(existingChildUser.userId, childUser);
        }
    }

    async getChildrenForParent(parentUsername: string): Promise<Array<User>> {
        const parentUser = await this.getUser(parentUsername);
        if (parentUser === undefined) {
            throw new UserNotFoundException(parentUsername);
        }
        console.log("getChildrenForParent(" + parentUsername + "), using parent ", parentUser);

        const children = this.userDataRepo.getChildrenForParent(parentUser.userId);
        return children;
    }

}

File: mgyoutube-multitech/api-go/src/modules/defaultusermoduleimpl.go
package modules

import (
	"apimodel"
	"errors"
	"log"
	"repos"
)

// DefaultUserModuleImpl = structure for default user module impl
type DefaultUserModuleImpl struct {
	userDataRepo repos.UserDataRepo
}

//AuthUser - authenticate the user
func (userModule DefaultUserModuleImpl) AuthUser(userCredential *apimodel.UserCredential) *apimodel.User {
	user := userModule.userDataRepo.GetUserByUsername(userCredential.Username)

	if user != nil {
		if len(user.Password) == 0 {
			log.Println("user.password is empty")
			user = nil
		} else if len(userCredential.Password) == 0 {
			log.Println("userCredential.password is empty")
			user = nil
		} else if user.Password != userCredential.Password {
			log.Println("user.password(", user.Password, ") != userCredential.password(", userCredential.Password, ")")
			user = nil
		} else {
			log.Println("password match!")
			// password match!
		}
	} else {
		log.Println("getUserByUsername(", userCredential.Username, ") returned None")
	}

	return user
}

// GetUserByUsername - returns the user for the username
func (userModule DefaultUserModuleImpl) GetUserByUsername(username string) *apimodel.User {
	return userModule.userDataRepo.GetUserByUsername(username)
}

// CreateUpdateUser - create or update based on if the username already exists
func (userModule *DefaultUserModuleImpl) CreateUpdateUser(user *apimodel.User) *apimodel.User {
	existingUser := userModule.userDataRepo.GetUserByUsername(user.Username)
	if existingUser != nil {
		if len(user.UserID) == 0 {
			user.UserID = existingUser.UserID
		}

		log.Println("createUpdateUser is updating existing user ", existingUser, " to be ", user)
		return userModule.UpdateUser(existingUser.UserID, user)
	}

	log.Println("createUpdateUser is creating new user ", user)
	return userModule.CreateUser(user)
}

// CreateUser - creates a user based on the provided model
func (userModule DefaultUserModuleImpl) CreateUser(user *apimodel.User) *apimodel.User {
	if len(user.UserID) == 0 {
		//log.Println("CreateUser, user", user)
		createdUser := userModule.userDataRepo.AddUser(user)
		//log.Println("CreateUser, createdUser", createdUser)
		return createdUser
	}
	//log.Println("CreateUser, returning nil")
	return nil
}

// UpdateUser - updates the associated user with the provided data
func (userModule DefaultUserModuleImpl) UpdateUser(userID string, user *apimodel.User) *apimodel.User {
	log.Println("DefaultUserModuleImpl.UpdateUser, userID", userID, "as user ", user)
	return userModule.userDataRepo.ReplaceUser(userID, user)
}

// RemoveUser - removes the identified user
func (userModule DefaultUserModuleImpl) RemoveUser(username string) *apimodel.User {
	user := userModule.GetUserByUsername(username)

	if user != nil && len(user.UserID) != 0 {
		userModule.userDataRepo.RemoveUser(user)
	}

	return user
}

// GetChildrenForParent - returns the children for the provided parent
func (userModule DefaultUserModuleImpl) GetChildrenForParent(parentUsername string) ([]*apimodel.User, error) {
	parentUser := userModule.GetUserByUsername(parentUsername)
	if parentUser == nil {
		return nil, errors.New("unable to find parent")
	}

	children := userModule.userDataRepo.GetChildrenForParent(parentUser.UserID)
	return children, nil
}

// AddUpdateChildToParent -
func (userModule DefaultUserModuleImpl) AddUpdateChildToParent(parentUsername string, childUser *apimodel.User) (*apimodel.User, error) {
	log.Println("DefaultUserModuleImpl.AddUpdateChildToParent, parentUsername", parentUsername, " childUser ", childUser)
	parentUser := userModule.GetUserByUsername(parentUsername)
	if parentUser == nil {
		return nil, errors.New("unable to find parent")
	}

	existingChildUser := userModule.GetUserByUsername(childUser.Username)
	if existingChildUser == nil {
		log.Println("DefaultUserModuleImpl.AddUpdateChildToParent, creating child ", childUser)

		createdChildUser := userModule.CreateUser(childUser)
		if createdChildUser == nil {
			log.Println("DefaultUserModuleImpl.AddUpdateChildToParent, creating child ", childUser, " failed ")
			return nil, errors.New("Creating child failed")
		}

		userModule.userDataRepo.AddChildToParent(parentUser.UserID, createdChildUser.UserID)
		return createdChildUser, nil
	}

	log.Println("DefaultUserModuleImpl.AddUpdateChildToParent, updating child ", childUser, " as ", existingChildUser.UserID)
	userModule.userDataRepo.AddChildToParent(parentUser.UserID, existingChildUser.UserID)
	updatedUser := userModule.UpdateUser(existingChildUser.UserID, childUser)
	return updatedUser, nil
}

// NewDefaultUserModuleImpl - create a new NewDefaultUserModuleImpl
func NewDefaultUserModuleImpl(userDataRepo repos.UserDataRepo) *DefaultUserModuleImpl {
	defUserModuleImpl := DefaultUserModuleImpl{}
	defUserModuleImpl.userDataRepo = userDataRepo
	return &defUserModuleImpl
}
